### PR TITLE
Fixed #27395 -- Updated sitemap to generate 'alternates'

### PR DESCRIPTION
**Pull request by fdemmer**
Updated the sitemap generator and default template to include
link elements with hreflang attribute to alternate language URLs.

https://code.djangoproject.com/ticket/27395

### CODE REVIEW

**Comment by charettes** [comment]
Hello @fdemmer, thank you for your contribution!

Tests for the new attributes should be added as well. Is that something you're planing to do?

**Comment by fdemmer** [comment]
> Tests for the new attributes should be added as well. Is that something you're planing to do?

will do...

**Comment by fdemmer** [comment]
Hi @charettes, thanks for the reminder, I have added tests!

File: django/contrib/sitemaps/templates/sitemap.xml [hunk 1]
```diff
@@ -1,13 +1,16 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">
```
Comment (carltongibson) [inline django/contrib/sitemaps/templates/sitemap.xml:2]: Can I ask you to explain this `xmlns:xhtml` change? No doubt it's correct, but it seems unrelated to the `hreflang` change — unless it's not?

Comment (carltongibson) [reply django/contrib/sitemaps/templates/sitemap.xml:2]: Update: this is as [advised by Google](https://support.google.com/webmasters/answer/189077)

**Comment by fdemmer** [comment]
Removed the ticket number from the comment; docs edits LGTM!

Thank you for digging this up! :+1:

**Comment by carltongibson** [comment]
Hi @fdemmer. Thanks for the follow-up. Glad you're still with us. 

We haven't had any comments from the list... — which maybe isn't important, but I'm just a little worried/concerned that we break something here in a way we didn't think about. Like maybe, generating the alternates doesn't work for a given site, because the `items` aren't properly localised or ... — I'm not using this myself so I'm not 100% sure of the ways it might go wrong. (This was why I posted to the list)

One question: if we were to add an `alternates` flag, defaulting to `False` could we maintain the existing behaviour behaviour entirely? (Then folks would opt in to the new feature.) How does that sound to you? Is there a case for not wanting the alternates generated? 

Thanks again! Lets try and come to a conclusion and move this forward one way or the other. 👍

**Comment by fdemmer** [comment]
Well, it _does_ change the output of the sitemap generator if `i18n` is set. One could easily argue, that that is a breaking change fit for a major release. Following that we should either delay until 4.0 and/or better yet make it optional.

Strictly following Google's recommendations there is no reason to _not_ have the alternates for a multi-language site, but it increases the load and of course the size of the response of the view.

I'll try to find time to add a switch like you mentioned over the weekend.

**Comment by carltongibson** [comment]
Hey @fdemmer. Thanks! (I think if the changes are purely opt-in then it's easier to merge as we don't have BC worries.)

File: docs/ref/contrib/sitemaps.txt [hunk 2]
```diff
@@ -252,6 +252,40 @@ Note:
...
+        generating alternate links when :attr:`~Sitemap.i18n` is enabled.
+        Defaults to :setting:`LANGUAGES`.
+
+    .. attribute:: Sitemap.alternates
+
+        .. versionadded:: 3.1
+
+        **Optional.**
+
+        When used in conjunction with :attr:`~Sitemap.i18n` enables that the
```
Comment (smithdc1) [inline docs/ref/contrib/sitemaps.txt:271]: This is a bit tricky to read. My suggestion would be to chop 'enables that the'

File: tests/sitemaps_tests/test_http.py [hunk 3]
```diff
@@ -259,19 +259,95 @@ def test_empty_sitemap(self):
 
     @override_settings(LANGUAGES=(('en', 'English'), ('pt', 'Portuguese')))
     def test_simple_i18nsitemap_index(self):
-        "A simple i18n sitemap index can be rendered"
+        """
+        A simple i18n sitemap index can be rendered.
+        """
```
Comment (smithdc1) [inline tests/sitemaps_tests/test_http.py:264]: Should there be a test that the variables are as expected (`i18n` `alternates` etc)?

Comment (fdemmer) [reply tests/sitemaps_tests/test_http.py:259]: Thanks for all the feedback! I'm not sure I understand this one; I followed the existing tests as examples.

The fact, that `i18n` and `alternates` are enabled can be deducted from the returned XML. Do you mean to validate the values of the attributes on eg. the `tests.sitemaps_tests.urls.http.AlternatesI18nSitemap` in the test fixture?

The `django.contrib.sitemaps.views.sitemap` view does not pass the `Sitemap` instance with those attributes to the template context. Is there another way to validate those from a response object?

Comment (smithdc1) [reply tests/sitemaps_tests/test_http.py:259]: Thanks for picking these up so quickly. :-)

I was trying to pick up on Simon's earlier comment. Maybe it's not needed or possible. 

I'd suggest unticking 'needs improvement' on trac now to get this back in the review queue?

Comment (fdemmer) [reply tests/sitemaps_tests/test_http.py:259]: > I'd suggest unticking 'needs improvement' on trac now to get this back in the review queue?

i did. let's get this done :)
